src/sinks/util/udp.rs

```diff
@@ -123,36 +141,62 @@ impl Into<UdpSink> for UdpConnector {
     }
 }
 
-impl Into<UdpService> for UdpConnector {
-    fn into(self) -> UdpService {
-        UdpService { connector: self }
-    }
+enum UdpServiceState {
+    Connecting(BoxFuture<'static, UdpSocket>),
+    Connected(UdpSocket),
 }
 
 pub struct UdpService {
-    connector: UdpConnector,
+    state: UdpServiceState,
+}
+
+impl Into<UdpService> for UdpConnector {
+    fn into(self) -> UdpService {
+        UdpService {
+            state: UdpServiceState::Connecting(self.connect_backoff().boxed()),
+        }
+    }
 }
 
 impl tower::Service<Bytes> for UdpService {
     type Response = ();
     type Error = UdpError;
-    type Future = BoxFuture<'static, Result<(), Self::Error>>;
-
-    fn poll_ready(&mut self, _: &mut Context<'_>) -> Poll<Result<(), Self::Error>> {
-        Poll::Ready(Ok(()))
+    type Future = future::Ready<Result<(), Self::Error>>;
+
+    fn poll_ready(&mut self, cx: &mut Context<'_>) -> Poll<Result<(), Self::Error>> {
+        match &mut self.state {
+            UdpServiceState::Connected(_) => Poll::Ready(Ok(())),
+            UdpServiceState::Connecting(fut) => match fut.poll_unpin(cx) {
+                Poll::Ready(socket) => {
+                    self.state = UdpServiceState::Connected(socket);
+                    Poll::Ready(Ok(()))
+                }
+                Poll::Pending => Poll::Pending,
+            },
```

Comment:
Realized that there is a shorter way :laughing: 
```rust
    fn poll_ready(&mut self, cx: &mut Context<'_>) -> Poll<Result<(), Self::Error>> {
        if let UdpServiceState::Connecting(ref mut fut) = self.state {
            let socket = futures::ready!(fut.poll_unpin(cx));
            self.state = UdpServiceState::Connected(socket);
        }
        Poll::Ready(Ok(()))
    }
```

Everything LGTM. Hope that will be merged soon :slightly_smiling_face: Otherwise a lot of conflicts in https://github.com/timberio/vector/pull/4597 in the future :disappointed: